### PR TITLE
[TI] Enable Rotating Device ID and EUI64 for SerialNumber

### PR DESCRIPTION
**Pull request by ti-tpan**
#### Problem

* Enable optional feature for Rotating Device ID.
* Enable use of device EUI64 for SerialNumber.

#### Change overview

* Updates to enable the items in "What's being fixed?" of this PR.
* Add wrapper function to read a Thread platform's IEEE EUI64 (for use as SerialNumber).
* Enable Rotating Device ID optional feature as part of BLE advertisements.

#### Testing

* Tested lock-app against chip-device-ctrl:
    * Commissioning
    * Cluster control

### CODE REVIEW

**Comment by ti-tpan** [comment]
FYI @srickardti @adabreuti

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (bzbarsky-apple) [inline examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:163]: Is there a reason not to use our existing BytesToHex utility here?

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (bzbarsky-apple) [inline examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:170]: Why the busy-loop?

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (bzbarsky-apple) [inline examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:164]: Why is this dependent on rotating device id bits?   The attribute is enabled unconditionally....

File: src/include/platform/ThreadStackManager.h
Comment (bzbarsky-apple) [inline src/include/platform/ThreadStackManager.h:100]: Please document (e.g. how big the buffer needs to be).

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (kkasperczyk-no) [inline src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1840]: Please be aware that using serial number for generating rotating device ID is not compatible with the spec, according to the:
> It SHALL NOT contain a fixed identifier such as a serial number
...
The unique identifier SHALL consist of a randomly-generated 128-bit or longer octet string which
SHALL be programmed during factory provisioning or delivered to the device by the vendor using
secure means after a software update.
The unique identifier SHALL be protected against reading or writing over the air after initial
introduction into the device, and stay fixed during the lifetime of the device.

 I know that other platforms are also using it and I'm trying to fix this in this PR: https://github.com/project-chip/connectedhomeip/pull/15031, but it there is still some refactoring needed before it will be merged.

File: examples/lock-app/cc13x2x7_26x2x7/args.gni
Comment (andy31415) [inline examples/lock-app/cc13x2x7_26x2x7/args.gni:38]: Most people will copy & paste examples, so our default was to disable rotating device id to make it a explicit user choice when compiling.

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (andy31415) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:170]: Maybe ChipDie would be more approrpiate. Or replace the if with `VerifyOrDie`

File: examples/lock-app/lock-common/lock-app.zap
Comment (andy31415) [inline examples/lock-app/lock-common/lock-app.zap:1840]: Why do we store a serial number in RAM? would a serial number not be fixed per device somewhere in flash?

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:163]: @bzbarsky-apple , thanks for pointing this out, BytesToHex does sound like the perfect API to use here. I did not know that existed, will try it out and update.

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:170]: Thanks for the suggestions, those sound like good candidates which would have similar effect as while loop, but much more intentional. This may also apply to our other busy-loops in this file.

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:164]: Correct, there is always a SerialNumber which is used (either one which has been stored in NV via KVS, or the hard-coded testing on "TEST_SN").

In this case, we are storing some other string as SerialNumber (vs "TEST_SN").

While this is not strictly dependent on rotating device id bits, when I worked on enabling this feature, it was in the context of enabling both CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING and CHIP_ENABLE_ROTATING_DEVICE_ID.

It is viable to store "stringEUI64" independently of these CHIP_ENABLE* bits, but I wanted to leave this up to the developer to decide. 
@bzbarsky-apple , do we think it is better practice to remove the CHIP_ENABLE* guards in this case?

File: src/include/platform/ThreadStackManager.h
Comment (ti-tpan) [reply src/include/platform/ThreadStackManager.h:100]: Of course, thank you for pointing this out!

File: examples/lock-app/cc13x2x7_26x2x7/args.gni
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/args.gni:38]: Makes sense, I can test this PR with rotating device id = true, but set it to false in the final PR.

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp]: Updated here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/763e4e38ae860e9f50e0b560c70b19833ef281c1

File: src/include/platform/ThreadStackManager.h
Comment (ti-tpan) [reply src/include/platform/ThreadStackManager.h:105]: Updated here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/8c702c8ab0a735f76a79de460b268a91687d4cf4

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp]: Updated here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/9bd267a4b72636f48a7990b51968bb575aaa1a01

File: examples/lock-app/cc13x2x7_26x2x7/args.gni
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/args.gni]: Updated here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/e5d41def8c65ec2f15371f497e5ce4e82ce4696c

File: examples/lock-app/lock-common/lock-app.zap
Comment (ti-tpan) [reply examples/lock-app/lock-common/lock-app.zap]: You're right, I mistakenly added this field -- I have removed it now here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/82a749ffe52646ba7422a171d29fa8f53d9b7184

The storageOption is indeed set to "external" for the Basic Cluster: https://github.com/ti-tpan/connectedhomeip/blob/ti/ffs/examples/lock-app/lock-common/lock-app.zap#L1292

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (ti-tpan) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1840]: @kkasperczyk-no , thanks for explaining.
Do you have an ETA on the merge? And followup, do you have a preference whether I merge this in before #15031 ?
- If merged before your PR, I can make a note that your PR must be considered before using the feature on TI platform.

(Since this feature is disabled by default (https://github.com/ti-tpan/connectedhomeip/blob/ti/ffs/examples/lock-app/cc13x2x7_26x2x7/args.gni#L39), it should at least (by default) be outside of this portion of spec).

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (bzbarsky-apple) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:163]: It's mostly a matter of what behavior we are aiming for here...  Either it makes sense to use the Thread EUI64 as the serial number, or it does not, and it doesn't seem like that should depend on the rotating device id bits...

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (bzbarsky-apple) [inline examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:166]:
```suggestion
			   "StoreSerialNumber: failed (%" CHIP_ERROR_FORMAT ")", ret.Format());
```
so this will work right if verbose error formatting is enabled.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (kkasperczyk-no) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1840]: I believe this is ready to be merged, as all comments were addressed and I don't see any other requests from some time. CI failures were also fixed. If it comes to the preferences I would love to merge my PR, as it is already open for 2 weeks and solving conflicts took me a lot of time, but in the end both solutions works for me.

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp]: Tested it out, and it works, thanks!
Updated here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/ae01057e16c3ce4e52e70d08b5f24f6784febecb

File: examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp
Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp:163]: Summary: I will remove these macro guards for now, as there seems to be no strict dependency on them.

The behavior I was aiming for, when these two macros are enabled, is to ensure a unique value for SerialNumber (IEEE 802.15.4 EUI64 should be unique per device).
When either macro is disabled, any value could be used -- I believe "TEST_SN" is the default. I wasn't sure whether any other test flow is assuming "TEST_SN", but I did know that at least Rotating Device ID feature is disabled by default (so results should be unchanged for any test flow which has rotating device ID disabled).

From my own simple tests of commissioning and cluster control, it seems that these two have no assumption of "TEST_SN" as SerialNumber, so I am happy to remove these macro guards. We can re-consider them in the future on an as-needed basis.

Comment (ti-tpan) [reply examples/lock-app/cc13x2x7_26x2x7/main/AppTask.cpp]: Update here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/63fa0edb1f455eecc5b58090d3593308e9e220e7

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (ti-tpan) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1840]: Sounds good, I will hold off on merging my PR here, and check back in few days on #15031 .

**Review by bzbarsky-apple — APPROVED** [review]

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (LuDuda) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1840]: @ti-tpan the #15031 has been already merged. Please align this PR. Thanks for waiting!

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (andy31415) [inline src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1836]: lifetimeCounter should be part of the `#if`. This means that `generateAdditionalDataPayload` should take some `AdditionalDataPayloadGeneratorParams` (which it seems to do according to current master: https://github.com/project-chip/connectedhomeip/blob/master/src/setup_payload/AdditionalDataPayloadGenerator.h#L79):


```cpp
CHIP_ERROR generateAdditionalDataPayload(AdditionalDataPayloadGeneratorParams & params,
                                             chip::System::PacketBufferHandle & bufferHandle,
                                             BitFlags<AdditionalDataFields> additionalDataFields);
```

Comment (andy31415) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1836]: It seems this needs merge with master and updates.

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (andy31415) [inline src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1833]: declare this where used not on top of the file. that reduces scope and makes this feel more C++ than C.

Comment (andy31415) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1833]: I see that you do this because `goto exit`. Could we not use goto-exit?

**Comment by woody-apple** [comment]
@ti-tpan Can you please rebase this?

**Comment by ti-tpan** [comment]
@woody-apple ,
Sorry, I have not had a chance to work on this PR in a while.
Let me try a quick re-base shortly.

My main concern with the PR is this comment: https://github.com/project-chip/connectedhomeip/pull/15367#discussion_r811117229
It seems I will need to re-work a bit.

File: src/platform/cc13x2_26x2/CC13X2_26X2Config.cpp
Comment (ti-tpan) [reply src/platform/cc13x2_26x2/CC13X2_26X2Config.cpp:152]: Good catch.

There is a chance `outLen == bufSize` (since [ReadConfigValueBin checks len <= bufSize](https://github.com/ti-tpan/connectedhomeip/blob/03f9947315d68210bd3dde571c3a410a8d9f9ff8/src/platform/cc13x2_26x2/CC13X2_26X2Config.cpp#L165)).

If a caller of `ReadConfigValueStr` does not account for additional null termination, then the null should not be added.

In a broader case, I'm not sure if _every_ caller of `ReadConfigValueStr` needs the null termination. If there is some call which does not require null termination, then we don't need to return the error.
As an example, [GetSerialNumber ](https://github.com/project-chip/connectedhomeip/blob/84c150a1b98746811b4931036d5ac3be0b2c592c/src/include/platform/internal/GenericConfigurationManagerImpl.cpp#L347)assumes null termination, and returns an error accordingly.
^ Does this seem redundant? Should the caller (in this case `GetSerialNumber`) handle `NULL` checks, or should `ReadConfigValueStr`?

So for now, I propose that `ReadConfigValueStr` performs NULL termination with the additional condition that `bufSize > outLen`.

Comment (ti-tpan) [reply src/platform/cc13x2_26x2/CC13X2_26X2Config.cpp:154]: Resolved here: https://github.com/project-chip/connectedhomeip/pull/15367/commits/a532260662e3863ce3dc67f5ec68a90a81a1b1af

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (ti-tpan) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1836]: Yes, you're right, this function call is stale.
I will address this on my next re-base (as the build error will show up).

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (ti-tpan) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:1833]: > declare this where used not on top of the function. that reduces scope and makes this feel more C++ than C.

Ok, we can move the variables to feel more like C++.
Specifically:
- we can move declaration of `packetBuf` to just before the `generateAdditionalDataPayload`
- we can move declaration of `len` to just before the `len =`

> I see that you do this because `goto exit`. Could we not use goto-exit?

What do you mean by `goto-exit`? How is this different from `goto exit` ?
My initial guess is that you mean to replace `SuccessOrExit(err);` with `goto exit` -- this way the reference to `err` is not needed when resolving macros.
So, I will do this for now, and if you mean differently, then I can re-consider it!

Comment (ti-tpan) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp]: Partially handled here: https://github.com/ti-tpan/connectedhomeip/commit/472f439e3ae0c0da2c18f4c59a1024af3d2d3daa

**Comment by ti-tpan** [comment]
Refresh stale-ness. To be updated.

**Comment by ti-tpan** [comment]
Refresh stale-ness. To be re-based and updated.

**Comment by andy31415** [comment]
@ti-tpan - is this still up to date? this seems to be open >1 month, so need to determine if to follow up or just close it.

**Comment by ti-tpan** [comment]
Hi @andy31415 , sorry, I was out of office. For now, let's close this. Based on my notes I should be looking back into this in 2-3 weeks.

Thanks

**Pull request closed without merging**
